Add support for chat streaming

dehoward opened the pull request:
### Motivation and Context

<!-- Thank you for your contribution to the copilot-chat repo!
Please help reviewers and future users, providing the following information:
  1. Why is this change required?
  2. What problem does it solve?
  3. What scenario does it contribute to?
  4. If it fixes an open issue, please link to the issue here.
-->
responses will now be streamed directly from the Kernel to all users in the chat. 
### Description
- `ChatSkill.cs`:
  -  added `StreamResponseToClient()` to broadcast the streamed response to all clients.
  - changed `GetChatResponseAsync()` to use `Task.WhenAll` when extracting audience/user intent and semantic/document memories.
  - changed `GetChatContextTokenLimit()` to account for the tokens used in extracting the audience

<!-- Describe your changes, the overall approach, the underlying design.
     These notes will help understanding how your code works. Thanks! -->

### Preview
**regular messages:**
![gif](https://github.com/microsoft/chat-copilot/assets/52973358/b5ba0b1f-3ddc-4648-8eff-587735770b7a)

**planner messages:**
![gif2](https://github.com/microsoft/chat-copilot/assets/52973358/79bfa394-7857-4e06-9302-174a4e6d1247)

**multiuser messages:**
![gif](https://github.com/microsoft/chat-copilot/assets/52973358/8013ce33-3eff-4bc2-bf74-dafac378e9b0)

### Contribution Checklist

<!-- Before submitting this PR, please make sure: -->

- [x] The code builds clean without any errors or warnings
- [x] The PR follows the [Contribution Guidelines](https://github.com/microsoft/copilot-chat/blob/main/CONTRIBUTING.md) and the [pre-submission formatting script](https://github.com/microsoft/copilot-chat/blob/main/CONTRIBUTING.md#dev-scripts) raises no violations
- [x] All unit tests pass, and I have added new tests where possible
- [x] I didn't break anyone :smile:

TaoChenOSU commented on webapi/CopilotChat/Skills/ChatSkills/ChatSkill.cs:
```diff
@@ -313,109 +294,112 @@ public async Task<string> ExtractAudienceAsync(SKContext context)
-        var userIntent = await this.GetUserIntentAsync(chatContext);
+        // Get the audience and user intent
+        await this.UpdateBotResponseStatusOnClient(chatId, "Extracting audience and user intent");
+        var tasks = await Task.WhenAll(this.GetAudienceAsync(chatContext), this.GetUserIntentAsync(chatContext));
```
Is this going to create a data race with the same chatContext? i.e one of the method results in error but the other one finishes later and override the error state.

dehoward replied:
reverted. the audience extraction code will no longer be async after #19

TaoChenOSU commented on webapi/CopilotChat/Skills/ChatSkills/ChatSkill.cs:
```diff
@@ -313,109 +294,112 @@ public async Task<string> ExtractAudienceAsync(SKContext context)
+            this._documentMemorySkill.QueryDocumentsAsync(userIntent, chatId, documentContextTokenLimit, chatContext.Memory)
+        );
+
         if (chatContext.ErrorOccurred)
```
This is not your code but we probably don't need this anymore as we are not passing the context to the `QueryXXXAsync` methods anymore. We need to throw and catch exceptions instead.

TaoChenOSU replied:
Can be addressed in a separate PR.

dehoward replied:
fixed

TaoChenOSU commented on webapi/CopilotChat/Skills/ChatSkills/ChatSkill.cs:
```diff
@@ -561,18 +545,19 @@ private async Task<ChatMessage> SaveNewMessageAsync(string message, string userI
+    /// <param name="userId">The user ID</param>
     /// <returns>The created chat message.</returns>
-    private async Task<ChatMessage> SaveNewResponseAsync(string response, string prompt, string chatId)
+    private async Task<ChatMessage?> SaveNewResponseAsync(string response, string prompt, string chatId, string userId)
```
Is `?` needed?

dehoward replied:
yes, since this function returns `null` when there is an error.

TaoChenOSU replied:
Which code path returns a null?

dehoward replied:
ah sorry you're right I got confused with `GetChatResponseAsync`. I've removed the nullable from this fn

TaoChenOSU reviewed (commented):
Have you tested the app with multiple users in the chat?

dehoward commented:
> Have you tested the app with multiple users in the chat?

yes, added a couple of needed tweaks in just now and also added a gif in the description of a multiuser example.

teresaqhoang commented on webapp/src/libs/services/ChatService.ts:
```diff
@@ -122,7 +121,7 @@ export class ChatService extends BaseService {
             ask.variables = ask.variables ? ask.variables.concat(openApiSkillVariables) : openApiSkillVariables;
         }
 
-        const result = await this.getResponseAsync<IAskResult>(
```
This is needed for token usages (coming in later PR) so req to leave as is

teresaqhoang commented on webapp/src/components/chat/ChatRoom.tsx:
```diff
@@ -65,7 +64,7 @@ export const ChatRoom: React.FC = () => {
     React.useEffect(() => {
         if (!shouldAutoScroll) return;
-        scrollToTarget(scrollTargetRef.current);
+        scrollViewTargetRef.current?.scrollTo(0, scrollViewTargetRef.current.scrollHeight);
```
nice!!

teresaqhoang commented on webapp/src/redux/features/message-relay/signalRMiddleware.ts:
```diff
@@ -146,34 +144,29 @@ export const signalRMiddleware = (store: StoreMiddlewareAPI) => {
     });
 
+    hubConnection.on(
+        SignalRCallbackMethods.ReceiveMessageContent,
```
nit: use `ReceiveMessageStream` instead of `ReceiveMessageContent` to be more indicative of content coming through, MessageContent and Message can be mistaken as synonymous

teresaqhoang reviewed (changes requested):
Copy nit (tagging @matthewbolanos if you have any input on the following): 

In the first screengrab gif, I recommend replacing `Generating bot response...` instead of `Invoking AI model` when the uber prompt is being invoked. 
- This helps keep the flow well defined in laymens terms, and we technically invoke the AI model on multiple occasions. 

And then using `Generating Semantic Memories` instead of `Extracting Semantic Memories...`.
- @TaoChenOSU and I had a conversation about this before and extraction gets a bit confusing since it makes it seem like we're on the step that pulling the memories to use for the response generation, but "generating" implies that this step creates new memories based on the chat history, rather than extracting them from somewhere else.

dehoward commented on webapp/src/components/chat/ChatInput.tsx:
```diff
@@ -14,10 +14,7 @@ import { GetResponseOptions, useChat } from '../../libs/useChat';
 import { useAppDispatch, useAppSelector } from '../../redux/app/hooks';
 import { RootState } from '../../redux/app/store';
 import { addAlert } from '../../redux/features/app/appSlice';
-import {
```
yes, fixed

dehoward commented:
> Copy nit (tagging @matthewbolanos if you have any input on the following):
> 
> In the first screengrab gif, I recommend replacing `Generating bot response...` instead of `Invoking AI model` when the uber prompt is being invoked.
> 
> * This helps keep the flow well defined in laymens terms, and we technically invoke the AI model on multiple occasions.
> 
> And then using `Generating Semantic Memories` instead of `Extracting Semantic Memories...`.
> 
> * @TaoChenOSU and I had a conversation about this before and extraction gets a bit confusing since it makes it seem like we're on the step that pulling the memories to use for the response generation, but "generating" implies that this step creates new memories based on the chat history, rather than extracting them from somewhere else.

good points. I've updated the strings to reflect this!

teresaqhoang commented on webapi/CopilotChat/Hubs/MessageRelayHub.cs:
```diff
@@ -39,17 +39,18 @@ public async Task AddClientToGroupAsync(string chatId)
     /// </summary>
-    /// <param name="chatId">The ChatID used as group id for SignalR.</param>
+    /// <param name="chatId">The chat ID used as group id for SignalR.</param>
+    /// <param name="userId">The user ID used as group id for SignalR.</param>
```
Why is both chatId and userId user for groupId here?
`userId` make more sense as `senderId` or `sourceId` since it's overloaded with the "userId" property that already exists in the ChatMessage object (This was confusing when trying to follow the CreateBotMessageOnClient flow in ChatSkill)

```suggestion
    /// <param name="userId">The user ID of the source of the message.</param>
```

teresaqhoang reviewed (changes requested):
Missed the ChatSkill file in my first pass, but thanks for adding all the comments in the ChatAsync method! A lot easier to follow now

dehoward commented on webapi/CopilotChat/Skills/ChatSkills/ChatSkill.cs:
```diff
@@ -313,109 +297,120 @@ public async Task<string> ExtractAudienceAsync(SKContext context)
     /// <param name="chatContext">The SKContext.</param>
-    /// <returns>A response from the model.</returns>
-    private async Task<string> GetChatResponseAsync(string chatId, SKContext chatContext)
+    /// <returns>The created chat message.</returns>
```
included them both
